orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -1895,5 +1895,71 @@ IMPLEMENT_GRADIENT_BUILDER(GetFakeQuantGradient) {
   return {NodeDef(OpDef{"FakeQuantGrad", kMSDomain, 1}, {GO(0), O(1)}, {GI(0)})};
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetLSTMGradient) {
+  std ::vector<ArgDef> input_args;
+
+  // Add inputs of the LSTMTraining node as inputs of the LSTMGrad node
+  // Add inputs from the source node for X, W, R, B, SL, H0, C0, P
+  // Add empty argdef for non existing inputs
+  const int input_size = GetSrcNodeInputSize();
+  for (int i = 0; i < 8; i++) {
+    if (i < input_size && I(i).Exists()) {
+      input_args.push_back(I(i));
+    } else {
+      input_args.push_back(ArgDef());
+    }
+  }
+
+  if (O(0).Exists()) {
+    input_args.push_back(O(0));  // all hidden states output of the LSTMTraining node
+  } else {
+    input_args.push_back(ArgDef());
+  }
+
+  if (O(3).Exists()) {
```

Comment:
do we need check output size before accessing O(3)?

Reply:
I can add an ORT_ENFORCE to make sure it exists. The transformer should ensure that this output exists.